hal/src/electron/modem/mdm_hal.cpp

```diff
@@ -1263,6 +1263,42 @@ bool MDMParser::powerState(void) const {
     return state;
 }
 
+bool MDMParser::softPowerOff(void) {
+    if (!powerState()) {
+        return true;
+    }
+    if (!(_init && _pwr)) {
+        MDM_INFO("\r\n[ Modem::softPowerOff ] Perform low level initialization to talk to the modem...");
+        uint8_t i = 0;
+        for (i = 0; i < 3; i++) {
+            if (_powerOn()) {
+                break;
+            }
+        }
+        if (i >= 3) {
+            // Failed to power on the modem.
+            MDM_INFO("\r\n[ Modem::softPowerOff ] Failed to power on the modem");
+            return false;
+        }
+    }
+    for (uint8_t i = 0; i < 3; i++) {
+        if (!_error && _atOk()) {
```

Comment:
Should we change this to `_checkModem(true)`?

Reply:
I think we can't. Since in `_checkModem()` it checks
```
if (_cancel_all_operations) {
        return false;
    }
```
 while `_cancel_all_operations` might be set after entering `softPowerOff()`, in that case we cannot continue the soft power off procedure.

Reply:
Should we at least drop `_error` check then perhaps?

Reply:
So we're going to power the modem off regardless of any error?

Reply:
We should probably at least try to. We still have the AT/OK check in there to make sure the modem is responsive.

Reply:
The logic is the same as is: https://github.com/particle-iot/device-os/blob/develop/hal/src/electron/modem/mdm_hal.cpp#L1283. If there is certain error that indicates that the mode is not accepting any AT command, we can probably retry or use PWR_ON going forward.

Reply:
No strong preferences, but we cannot use `PWR_ON` on G350, so the suggestion is to ignore any errors that happened previously and just rely on AT/OK, as otherwise there is no reason to also loop at all once we see `_error`, as there is nothing in the loop that can potentially reset it here in place.

Reply:
Makes sense. Will drop the `_error` check.